crates/ruff/src/rules/ruff/rules/unreachable.rs

```diff
@@ -0,0 +1,964 @@
+                error!("Got an unexpected empty code block");
+                return None;
+            };
+            Some(Diagnostic::new(
+                UnreachableCode {
+                    name: name.as_str().to_owned(),
+                },
+                stmt.range(),
+            ))
+        })
+        .collect()
+}
+
+/// Simple bitmap.
+#[derive(Debug)]
+struct Bitmap {
+    bits: Box<[usize]>,
+    capacity: usize,
+}
+
+impl Bitmap {
+    /// Create a new `Bitmap` with `capacity` capacity.
+    fn with_capacity(capacity: usize) -> Bitmap {
+        let mut size = capacity / usize::BITS as usize;
+        if (capacity % usize::BITS as usize) != 0 {
+            size += 1;
+        }
+        Bitmap {
+            bits: vec![0; size].into_boxed_slice(),
+            capacity,
+        }
+    }
+
+    /// Set bit at index `idx` to true.
+    ///
+    /// Returns a boolean indicating if the bit was already set.
+    fn set(&mut self, idx: usize) -> bool {
+        let n = idx / usize::BITS as usize;
+        let s = idx % usize::BITS as usize;
+        if (self.bits[n] & (1 << s)) == 0 {
+            self.bits[n] |= 1 << s;
+            false
+        } else {
+            true
+        }
+    }
+
+    /// Returns an iterator of all unset indices.
+    fn unset(&self) -> impl Iterator<Item = usize> + '_ {
+        let mut index = 0;
+        let mut shift = 0;
+        let last_max_shift = self.capacity % usize::BITS as usize;
+        iter::from_fn(move || loop {
+            if shift >= usize::BITS as usize {
+                shift = 0;
+                index += 1;
+            }
+            if self.bits.len() <= index || (index >= self.bits.len() - 1 && shift >= last_max_shift)
+            {
+                return None;
+            }
+
+            let is_set = (self.bits[index] & (1 << shift)) != 0;
+            shift += 1;
+            if !is_set {
+                return Some((index * usize::BITS as usize) + shift - 1);
+            }
+        })
+    }
+}
+
+/// Set bits in `reached_map` for all blocks that are reached in `blocks`
+/// starting with block at index `idx`.
+fn mark_reached(reached_map: &mut Bitmap, blocks: &[BasicBlock<'_>], mut idx: BlockIndex) {
+    loop {
+        let block = &blocks[idx];
+        if reached_map.set(idx) {
+            return; // Block already visited, no needed to do it again.
+        }
+
+        match &block.next {
+            NextBlock::Always(next) => idx = *next,
+            NextBlock::If {
+                condition,
+                next,
+                orelse,
+            } => {
+                match taken(condition) {
+                    Some(true) => idx = *next,    // Always taken.
+                    Some(false) => idx = *orelse, // Never taken.
+                    None => {
+                        // Don't know, both branches might be taken.
+                        idx = *next;
+                        mark_reached(reached_map, blocks, *orelse);
```

Comment:
What are your thoughts on using a Vec containing the blocks that need visiting vs using recursion?

Reply:
That would work as well. I mainly chose recursion for simplicity and I didn't want to allocate for a single index, which might be common for small functions. If we go with the `vec` route we could think about something like `smallvec` (https://crates.io/crates/smallvec) to avoid small heap allocations.